### PR TITLE
Fix service typings

### LINKED ISSUE #14035: Typescript service typing doesn't allow addition of custom services
### Required System information

- Strapi version: `4.3.2`

### Describe the bug

Typescript service typing doesn't allow addition of custom services, when adding a new service such as follows:

```ts
import { factories } from '@strapi/strapi'; 

export default factories.createCoreService('api::something.something', ({strapi}) =>  ({
  // Method 1: Creating an entirely custom service
  async exampleService(...args) {
    let response = { okay: true }

    if (response.okay === false) {
      return { response, error: true }
    }

    return response
  },
}));
```
It throws a typing error as follows:

image

### Expected behavior

Should work and not throw an error

### PR DESCRIPTION
**Pull request by Bassel17**
### What does it do?

- Adds a `GenericService` type that allows the creation of custom services

### How to test it?

- Try to add a custom service it should work just fine and not throw an error

### Related issue(s)/PR(s)

Fixes #14035

### CODE REVIEW

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
You will also want to change the service declaration in packages/core/strapi/lib/global.d.ts :)

**Comment by pwizla** [comment]
Hi! I guess this would not require any documentation update, it will just fix things, so they work as documented in [Backend customization > Services](https://docs.strapi.io/developer-docs/latest/development/backend-customization/services.html#adding-a-new-service), right?

File: packages/core/strapi/lib/types/core/strapi/index.d.ts [hunk 1]
```diff
@@ -33,12 +34,12 @@ export interface Strapi {
...
    * It returns all the registered services
    */
-  readonly services: StringMap<Service>;
+  readonly services: StringMap<GenericService>;
 
   /**
    * Find a service using its unique identifier
    */
-  service<T extends Service = unknown>(uid: string): T | undefined;
+  service(uid: string): GenericService | undefined;
```
Comment (innerdvations) [inline packages/core/strapi/lib/types/core/strapi/index.d.ts:42]: Would something like this help to allow a specific type when you know what you're looking for?

```suggestion
  service<T extends GenericService = GenericService>(uid: string): T | undefined;
```

Comment (Bassel17) [reply packages/core/strapi/lib/types/core/strapi/index.d.ts:42]: yep I guess it would help

**Review by innerdvations — APPROVED** [review]

**Review by innerdvations — DISMISSED** [review]

File: packages/core/strapi/lib/core-api/service/index.d.ts [hunk 2]
```diff
@@ -20,3 +20,8 @@ export interface CollectionTypeService extends BaseService {
 
 export type Service = SingleTypeService | CollectionTypeService;
 
+export type GenericService = Partial<Service> & {
+  [method: string | number | symbol]: (...args:unknown) => Promise<Entity> | Entity;
```
Comment (Convly) [inline packages/core/strapi/lib/core-api/service/index.d.ts:24]: It won't necessarily return an Entity here :slightly_smiling_face:

**Review by Convly — CHANGES_REQUESTED** [review]

File: packages/core/strapi/lib/core-api/service/index.d.ts (hunk 2 above)
Comment (Bassel17) [reply packages/core/strapi/lib/core-api/service/index.d.ts:24]: hmmmm, should we have it be `any` or `unknown` ?

Comment (stafyniaksacha) [reply packages/core/strapi/lib/core-api/service/index.d.ts:24]: Can use a generic here:
`[method: string | number | symbol]: <T = unknown>(...args:unknown) => T;`

Comment (Bassel17) [reply packages/core/strapi/lib/core-api/service/index.d.ts:24]: @stafyniaksacha agreed probably using any would be better for now, will go with your suggestion

**Comment by eucciferri** [comment]
Hi @Bassel17! How are you?

I updated to the latest version and all my methods now throw me an error:

image

I looked for the definitions file and I modified the return type of the methods as follows:

image

After doing this everything works perfectly, even the custom services methods (which was the reason that generated this fix).

Is it possible for you to modify the return type?

Thanks!

**Comment by eucciferri** [comment]
Perhaps a return type like the following is more comprehensive:

```
export type GenericService = Partial<Service> & {
   [method: string | number | symbol]: <T = any>(...args: any) => Promise<Entity | Entity[] | void> | Entity | Entity[] | void;
};
```

**Comment by Bassel17** [comment]
@eucciferri it was already making an issue before which is why we made the return value into "any" or whatever the person chooses to type it, I'll check again

**Comment by stafyniaksacha** [comment]
> Is it possible for you to modify the return type?

No the error is good! It says that you have miss typed the findOne method: You used a string type whereas it needs a number

**Comment by stafyniaksacha** [comment]
For now to properly infer types while using `createCoreService`, this is what I use:

```ts
import type {
  CollectionTypeService,
  Service,
} from '@strapi/strapi/lib/core-api/service';

import { Strapi, factories } from '@strapi/strapi';

const custom = ({ strapi }: { strapi: Strapi }) => ({
  async foo() { return 'a string' }
  async bar() { return 42 }
})

const base = factories.createCoreService(
  'plugin::shopify.fulfillment-order',
  custom as Service
);

// this allow typescript to infer `foo` & `bar` functions with base `find`/`findOne`/....
export default base as () => CollectionTypeService & ReturnType<typeof custom>;
```

**Comment by stafyniaksacha** [comment]
Here is a shims for dynamic `createCoreService` method:

```ts
import {
  Service,
  SingleTypeService,
  CollectionTypeService,
} from '@strapi/strapi/lib/core-api/service';

declare module '@strapi/strapi/lib/types/factories' {
  export interface ServiceUidMap {
    'plugin::foo.bar': CollectionTypeService; 
    'api::homepage.homepage': SingleTypeService; 
    [key: string]: Service;
  }

  export function createCoreService<K extends keyof ServiceUidMap, T = any>(
    uid: K,
    customActions?: T
  ): T extends (...args: any) => any
    ? () => ReturnType<T> & ServiceUidMap[K]
    : () => T & ServiceUidMap[K];
}

export {};

```

**Comment by eucciferri** [comment]
> No the error is good! It says that you have miss typed the findOne method: You used a string type whereas it needs a number for the entityId

Hi @stafyniaksacha! How are you? 

The error is reproduced regardless of the type of data that I place in the attributes, even if it is a "string", "number" or "any".

The only solution I found was to specify the type of response that the method can throw, so I placed a Promise (in case an async method is created) of an Entity, an array of Entities or a void (in case nothing is returned).

This error is always reproduced, even if you create a project from scratch and follow all the steps as indicated in the documentation, to create a new method or overwrite an existing one.

image

[https://docs.strapi.io/developer-docs/latest/development/backend-customization/services.html#adding-a-new-service](url)

Thanks.

**Comment by emahuni** [comment]
The error is still happening, however, doing the following fixes it I have no idea if this is desirable or not, coz I think it's up to the to decide the return type:

putting `any` as the return type completely fixed the issue such that everything works without a problem:

```ts

export type GenericService = Partial<Service> & {
  [method: string | number | symbol]: <T = any>(...args: any) => any;
};


export default factories.createCoreService('api::something.something', ({ strapi })  => {
  return {
    // service functions or properties here
    async someFunction (a: number): Promise<number[]> {
      return [1, 2, 3];
    },
    
    async someFunction2<T> (a: number): Promise<T[]> {
      return [1 as T]; // to show example
    },
  };
});

```

notice the generic example at the end still works, so I am assuming that anything that is returned should work.

**Comment by eucciferri** [comment]
Hi @emahuni, yes, it keeps getting the error for me too. I had to migrate all my services files to JavaScript in order to resolve this error.

@stafyniaksacha @Bassel17 Hi guys! Could you please review this issue?

**Comment by stafyniaksacha** [comment]
Hello @eucciferri! 

I created two PR here: 
- https://github.com/strapi/strapi/pull/14222
- https://github.com/strapi/strapi/pull/14219

**Comment by eucciferri** [comment]
Great news @stafyniaksacha! Thank you so much!! 😃

**Pull request merged**